File: web/e-platform-web/src/services/AdminService.js
// src/services/AdminService.js
const API_URL = process.env.REACT_APP_BFF_API_URL || 'https://tunsiaed.onrender.com';

const getAuthHeaders = (token) => ({
  'Authorization': `Bearer ${token}`,
  'Content-Type': 'application/json',
});

class AdminService {
  /**
   * Fetches admin dashboard statistics
   * @param {string} token - Authentication token
   * @returns {Promise<Object>} Admin stats data
   */
  static async getStats(token) {
    const res = await fetch(`${API_URL}/api/v1/admin/stats`, { headers: getAuthHeaders(token) });
    if (!res.ok) throw new Error('Failed to fetch stats');
    return res.json();
  }

  /**
   * Fetches revenue analytics data
   * @param {string} token - Authentication token
   * @returns {Promise<Object>} Revenue data
   */
  static async getRevenueData(token) {
    const res = await fetch(`${API_URL}/api/v1/admin/revenue`, { headers: getAuthHeaders(token) });
    if (!res.ok) throw new Error('Failed to fetch revenue');
    return res.json();
  }

  /**
   * Fetches recent platform activity
   * @param {string} token - Authentication token
   * @returns {Promise<Array>} Recent activity logs
   */
  static async getRecentActivity(token) {
    const res = await fetch(`${API_URL}/api/v1/admin/activity`, { headers: getAuthHeaders(token) });
    if (!res.ok) throw new Error('Failed to fetch activity');
    return res.json();
  }

  /**
   * Fetches course performance metrics
   * @param {string} token - Authentication token
   * @returns {Promise<Array>} Course performance data
   */
  static async getCoursePerformance(token) {
    const res = await fetch(`${API_URL}/api/v1/admin/courses/performance`, { headers: getAuthHeaders(token) });
    if (!res.ok) throw new Error('Failed to fetch course performance');
    return res.json();
  }

  /**
   * Fetches user engagement metrics
   * @param {string} token - Authentication token
   * @returns {Promise<Object>} User engagement data
   */
  static async getUserEngagement(token) {
    const res = await fetch(`${API_URL}/api/v1/admin/engagement`, { headers: getAuthHeaders(token) });
    if (!res.ok) throw new Error('Failed to fetch user engagement');
    return res.json();
  }

  /**
   * Fetches active promotions
   * @param {string} token - Authentication token
   * @returns {Promise<Array>} List of active promotions
   */
  static async getActivePromotions(token) {
    const res = await fetch(`${API_URL}/api/v1/admin/promotions/active`, { headers: getAuthHeaders(token) });
    if (!res.ok) throw new Error('Failed to fetch active promotions');
    return res.json();
  }

  /**
   * Fetches subscription plans
   * @param {string} token - Authentication token
   * @returns {Promise<Array>} List of subscription plans
   */
  static async getSubscriptionPlans(token) {
    const res = await fetch(`${API_URL}/api/v1/admin/subscriptions/plans`, { headers: getAuthHeaders(token) });
    if (!res.ok) throw new Error('Failed to fetch subscription plans');
    return res.json();
  }

  /**
   * Fetches subscription statistics
   * @param {string} token - Authentication token
   * @returns {Promise<Object>} Subscription stats
   */
  static async getSubscriptionStats(token) {
    const res = await fetch(`${API_URL}/api/v1/admin/subscriptions/stats`, { headers: getAuthHeaders(token) });
    if (!res.ok) throw new Error('Failed to fetch subscription stats');
    return res.json();
  }

  /**
   * Updates a subscription plan
   * @param {string} token - Authentication token
   * @param {string} planId - ID of the plan to update
   * @param {Object} data - Updated plan data
   * @returns {Promise<Object>} Updated plan data
   */
  static async updateSubscriptionPlan(token, planId, data) {
    const res = await fetch(`${API_URL}/api/v1/admin/subscriptions/${planId}`, {
      method: 'PATCH',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify(data),
    });
    if (!res.ok) throw new Error('Failed to update subscription plan');
    return res.json();
  }

  /**
   * Creates a new promotion
   * @param {string} token - Authentication token
   * @param {Object} data - Promotion data
   * @returns {Promise<Object>} Created promotion data
   */
  static async createPromotion(token, data) {
    const res = await fetch(`${API_URL}/api/v1/admin/promotions`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify(data),
    });
    if (!res.ok) throw new Error('Failed to create promotion');
    return res.json();
  }

  /**
   * Get all users (admin only)
   * @param {string} token - Authentication token
   * @param {Object} options - Query options (page, limit, role)
   * @returns {Promise<Array>} List of users
   */
  static async getAllUsers(token, options = {}) {
    const params = new URLSearchParams();
    if (options.page) params.append('page', options.page);
    if (options.limit) params.append('limit', options.limit);
    if (options.role) params.append('role', options.role);
    
    const res = await fetch(`${API_URL}/api/v1/admin/users?${params}`, {
      headers: getAuthHeaders(token),
    });
    if (!res.ok) throw new Error('Failed to fetch users');
    return res.json();
  }

  /**
   * Ban user (admin only)
   * @param {string} token - Authentication token
   * @param {string} userId - User ID to ban
   * @returns {Promise<Object>} Result message
   */
  static async banUser(token, userId) {
    const res = await fetch(`${API_URL}/api/v1/admin/users/${userId}/ban`, {
      method: 'PATCH',
      headers: getAuthHeaders(token),
    });
    if (!res.ok) throw new Error('Failed to ban user');
    return res.json();
  }

  /**
   * Unban user (admin only)
   * @param {string} token - Authentication token
   * @param {string} userId - User ID to unban
   * @returns {Promise<Object>} Result message
   */
  static async unbanUser(token, userId) {
    const res = await fetch(`${API_URL}/api/v1/admin/users/${userId}/unban`, {
      method: 'PATCH',
      headers: getAuthHeaders(token),
    });
    if (!res.ok) throw new Error('Failed to unban user');
    return res.json();
  }

  /**
   * Approve instructor application (admin only)
   * @param {string} token - Authentication token
   * @param {string} userId - User ID to approve as instructor
   * @returns {Promise<Object>} Result message
   */
  static async approveInstructor(token, userId) {
    const res = await fetch(`${API_URL}/api/v1/admin/instructors/${userId}/approve`, {
      method: 'PATCH',
      headers: getAuthHeaders(token),
    });
    if (!res.ok) throw new Error('Failed to approve instructor');
    return res.json();
  }

  /**
   * Decline instructor application (admin only)
   * @param {string} token - Authentication token
   * @param {string} userId - User ID to decline
   * @param {string} reason - Decline reason
   * @returns {Promise<Object>} Result message
   */
  static async declineInstructor(token, userId, reason) {
    const res = await fetch(`${API_URL}/api/v1/admin/instructors/${userId}/decline`, {
      method: 'PATCH',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify({ reason }),
    });
    if (!res.ok) throw new Error('Failed to decline instructor');
    return res.json();
  }

  /**
   * Get all courses (admin only)
   * @param {string} token - Authentication token
   * @param {Object} options - Query options (page, limit, status)
   * @returns {Promise<Array>} List of courses
   */
  static async getAllCourses(token, options = {}) {
    const params = new URLSearchParams();
    if (options.page) params.append('page', options.page);
    if (options.limit) params.append('limit', options.limit);
    if (options.status) params.append('status', options.status);
    
    const res = await fetch(`${API_URL}/api/v1/admin/courses?${params}`, {
      headers: getAuthHeaders(token),
    });
    if (!res.ok) throw new Error('Failed to fetch courses');
    return res.json();
  }

  /**
   * Approve course (admin only)
   * @param {string} token - Authentication token
   * @param {string} courseId - Course ID to approve
   * @returns {Promise<Object>} Result message
   */
  static async approveCourse(token, courseId) {
    const res = await fetch(`${API_URL}/api/v1/admin/courses/${courseId}/approve`, {
      method: 'PATCH',
      headers: getAuthHeaders(token),
    });
    if (!res.ok) throw new Error('Failed to approve course');
    return res.json();
  }

  /**
   * Reject course (admin only)
   * @param {string} token - Authentication token
   * @param {string} courseId - Course ID to reject
   * @param {string} reason - Rejection reason
   * @returns {Promise<Object>} Result message
   */
  static async rejectCourse(token, courseId, reason) {
    const res = await fetch(`${API_URL}/api/v1/admin/courses/${courseId}/reject`, {
      method: 'PATCH',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify({ reason }),
    });
    if (!res.ok) throw new Error('Failed to reject course');
    return res.json();
  }
}

export default AdminService;

File: mobile/e_platform_mobile/lib/Service/admin/admin_service.dart
import 'package:dio/dio.dart';

import '../../core/api/api_client.dart';
import '../../core/api/api_endpoints.dart';

/// Admin Service
/// Handles all admin-related API calls
class AdminService {
  final ApiClient _apiClient = ApiClient();

  /// Get admin dashboard
  Future<Map<String, dynamic>> getAdminDashboard() async {
    Map<String, dynamic> _mapOrEmpty(dynamic source) {
      if (source is Map<String, dynamic>) {
        return source;
      }
      if (source is Map) {
        return source.map((key, value) => MapEntry(key.toString(), value));
      }
      return <String, dynamic>{};
    }

    List<dynamic> _listOrEmpty(dynamic source) {
      if (source is List) {
        return source;
      }
      return const <dynamic>[];
    }

    final responses = await Future.wait<Response<dynamic>>(
      [
        _apiClient.get(ApiEndpoints.adminStats),
        _apiClient.get(ApiEndpoints.adminRevenue),
        _apiClient.get(ApiEndpoints.adminActivity),
        _apiClient.get(ApiEndpoints.adminCoursePerformance),
        _apiClient.get(ApiEndpoints.adminUserEngagement),
        _apiClient.get(ApiEndpoints.adminActivePromotions),
        _apiClient.get(ApiEndpoints.adminSubscriptionPlans),
        _apiClient.get(ApiEndpoints.adminSubscriptionStats),
      ],
      eagerError: true,
    );

    return {
      'stats': _mapOrEmpty(responses[0].data),
      'revenueChart': _mapOrEmpty(responses[1].data),
      'recentActivity': _listOrEmpty(responses[2].data),
      'coursePerformance': _listOrEmpty(responses[3].data),
      'userEngagement': _mapOrEmpty(responses[4].data),
      'activePromotions': _listOrEmpty(responses[5].data),
      'subscriptionPlans': _listOrEmpty(responses[6].data),
      'subscriptionStats': _mapOrEmpty(responses[7].data),
    };
  }

  /// Get admin stats
  Future<Map<String, dynamic>> getAdminStats() async {
    final response = await _apiClient.get(
      ApiEndpoints.adminStats,
    );
    return response.data;
  }

  /// Get all users (admin view)
  Future<List<Map<String, dynamic>>> getAllUsers({
    int? page,
    int? limit,
    String? role,
  }) async {
    final response = await _apiClient.get(
      ApiEndpoints.adminUsers,
      queryParameters: {
        if (page != null) 'page': page,
        if (limit != null) 'limit': limit,
        if (role != null) 'role': role,
      },
    );
    return List<Map<String, dynamic>>.from(response.data);
  }

  /// Get all courses (admin view)
  Future<List<Map<String, dynamic>>> getAllCourses({
    int? page,
    int? limit,
    String? status,
  }) async {
    final response = await _apiClient.get(
      ApiEndpoints.adminCourses,
      queryParameters: {
        if (page != null) 'page': page,
        if (limit != null) 'limit': limit,
        if (status != null) 'status': status,
      },
    );
    return List<Map<String, dynamic>>.from(response.data);
  }

  /// Approve course
  Future<Map<String, dynamic>> approveCourse(String courseId) async {
    final response = await _apiClient.patch(
      '${ApiEndpoints.adminCourses}/$courseId/approve',
    );
    return response.data;
  }

  /// Reject course
  Future<Map<String, dynamic>> rejectCourse(
    String courseId,
    String reason,
  ) async {
    final response = await _apiClient.patch(
      '${ApiEndpoints.adminCourses}/$courseId/reject',
      data: {'reason': reason},
    );
    return response.data;
  }

  /// Ban user
  Future<Map<String, dynamic>> banUser(String userId) async {
    final response = await _apiClient.patch(
      '${ApiEndpoints.adminUsers}/$userId/ban',
    );
    return response.data;
  }

  /// Unban user
  Future<Map<String, dynamic>> unbanUser(String userId) async {
    final response = await _apiClient.patch(
      '${ApiEndpoints.adminUsers}/$userId/unban',
    );
    return response.data;
  }

  /// Get revenue analytics (standardized method name)
  Future<Map<String, dynamic>> getRevenueData() async {
    final response = await _apiClient.get(
      ApiEndpoints.adminRevenue,
    );
    return response.data;
  }

  /// Get recent platform activity
  Future<List<Map<String, dynamic>>> getRecentActivity({int limit = 10}) async {
    final response = await _apiClient.get(
      ApiEndpoints.adminActivity,
      queryParameters: {
        'limit': limit,
      },
    );
    return List<Map<String, dynamic>>.from(response.data);
  }

  /// Legacy alias for backward compatibility
  @Deprecated('Use getRevenueData() instead')
  Future<Map<String, dynamic>> getRevenue() => getRevenueData();

  /// Get course performance metrics
  Future<List<Map<String, dynamic>>> getCoursePerformance() async {
    final response = await _apiClient.get(
      ApiEndpoints.adminCoursePerformance,
    );
    return List<Map<String, dynamic>>.from(response.data);
  }

  /// Get user engagement metrics
  Future<Map<String, dynamic>> getUserEngagement() async {
    final response = await _apiClient.get(
      ApiEndpoints.adminUserEngagement,
    );
    return response.data;
  }

  /// Get active promotions
  Future<List<Map<String, dynamic>>> getActivePromotions() async {
    final response = await _apiClient.get(
      ApiEndpoints.adminActivePromotions,
    );
    return List<Map<String, dynamic>>.from(response.data);
  }

  /// Create a new promotion
  Future<Map<String, dynamic>> createPromotion(
    Map<String, dynamic> data,
  ) async {
    final response = await _apiClient.post(
      ApiEndpoints.adminPromotions,
      data: data,
    );
    return response.data;
  }

  /// Get subscription plans
  Future<List<Map<String, dynamic>>> getSubscriptionPlans() async {
    final response = await _apiClient.get(
      ApiEndpoints.adminSubscriptionPlans,
    );
    return List<Map<String, dynamic>>.from(response.data);
  }

  /// Get subscription statistics
  Future<Map<String, dynamic>> getSubscriptionStats() async {
    final response = await _apiClient.get(
      ApiEndpoints.adminSubscriptionStats,
    );
    return response.data;
  }

  /// Update subscription plan
  Future<Map<String, dynamic>> updateSubscriptionPlan(
    String planId,
    Map<String, dynamic> data,
  ) async {
    final response = await _apiClient.patch(
      ApiEndpoints.adminSubscriptionPlan(planId),
      data: data,
    );
    return response.data;
  }
}
